x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/hooks/use_columns.ts

```diff
@@ -0,0 +1,228 @@
+  id,
+  defaultColumns,
+}: UseColumnsArgs) => {
+  const [isBrowserFieldDataLoading, browserFields] = useFetchBrowserFieldCapabilities({
+    featureIds,
+  });
+  const [columns, setColumns] = useState<EuiDataGridColumn[]>(storageAlertsTable.current.columns);
+  const [isColumnsPopulated, setColumnsPopulated] = useState<boolean>(false);
+  const [visibleColumns, setVisibleColumns] = useState(
+    storageAlertsTable.current.visibleColumns ?? []
+  );
+
+  useEffect(() => {
+    if (isBrowserFieldDataLoading !== false || isColumnsPopulated === true) return;
+
+    const populatedColumns = populateColumns(columns, browserFields);
+    setColumnsPopulated(true);
+    setColumns(populatedColumns);
+  }, [browserFields, columns, isBrowserFieldDataLoading, isColumnsPopulated]);
+
+  const onColumnsChange = useCallback(
+    (newColumns: EuiDataGridColumn[], newVisibleColumns: string[]) => {
+      setColumns(newColumns);
+      persist({
+        id,
+        storage,
+        storageAlertsTable,
+        columns: newColumns,
+        visibleColumns: newVisibleColumns,
+      });
+    },
+    [id, storage, storageAlertsTable]
+  );
+
+  const onChangeVisibleColumns = useCallback(
+    (newColumns: string[]) => {
+      setVisibleColumns(newColumns);
+      onColumnsChange(
+        columns.sort((a, b) => newColumns.indexOf(a.id) - newColumns.indexOf(b.id)),
+        newColumns
+      );
+    },
+    [onColumnsChange, columns]
+  );
+
+  const onToggleColumn = useCallback(
+    (columnId: string): void => {
+      const visibleIndex = visibleColumns.indexOf(columnId);
+      const defaultIndex = defaultColumns.findIndex(
+        (column: EuiDataGridColumn) => column.id === columnId
+      );
+
+      const isVisible = visibleIndex >= 0;
+      const isInDefaultConfig = defaultIndex >= 0;
+
+      let newColumnIds: string[] = [];
+
+      // if the column is shown, remove it
+      if (isVisible) {
+        newColumnIds = [
+          ...visibleColumns.slice(0, visibleIndex),
+          ...visibleColumns.slice(visibleIndex + 1),
+        ];
+      }
+
+      // if the column isn't shown but it's part of the default config
+      // insert into the same position as in the default config
+      if (!isVisible && isInDefaultConfig) {
+        newColumnIds = [
+          ...visibleColumns.slice(0, defaultIndex),
+          columnId,
+          ...visibleColumns.slice(defaultIndex),
+        ];
+      }
+
+      // if the column isn't shown and it's not part of the default config
+      // push it into the last position
+      if (!isVisible && !isInDefaultConfig) {
+        newColumnIds = [...visibleColumns, columnId];
+      }
+
+      const newColumns = newColumnIds.map((_columnId) => {
+        const column = getColumnByColumnId(defaultColumns, _columnId);
+
+        return euiColumnFactory(column ? column : { id: _columnId }, browserFields);
+      });
+
+      onChangeVisibleColumns(newColumnIds);
+      onColumnsChange(newColumns, newColumnIds);
```

Comment:
This will cause 2 `columns` state changes and 2 `persist` calls. Maybe we could move the `setVisibleColumns(newVisibleColumns);` line from the `onChangeVisibleColumns` to the `onColumnsChange`, and call only this one here? 🤔

Reply:
I'll call the setter functions instead of the callbacks